drivers/ipm/ipm_sedi.c

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2020-2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT intel_sedi_ipm
+
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/device.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/sys_io.h>
+#include <errno.h>
+
+#include <zephyr/devicetree.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ipm_sedi, CONFIG_IPM_LOG_LEVEL);
+
+#ifdef CONFIG_SHARED_IRQ
+#include <zephyr/shared_irq.h>
+#endif
+
+#include <zephyr/drivers/ipm.h>
+#include "ipm_sedi.h"
+
+extern void sedi_ipc_isr(IN sedi_ipc_t ipc_device);
+
+static void set_ipm_dev_busy(const struct device *dev, bool is_write)
+{
+	struct ipm_sedi_context *ipm = dev->data;
+	unsigned int key = irq_lock();
+
+	atomic_set_bit(&ipm->status, is_write ? IPM_WRITE_BUSY_BIT : IPM_READ_BUSY_BIT);
+	pm_device_busy_set(dev);
+	irq_unlock(key);
+}
+
+static void clear_ipm_dev_busy(const struct device *dev, bool is_write)
+{
+	struct ipm_sedi_context *ipm = dev->data;
+	unsigned int key = irq_lock();
+
+	atomic_clear_bit(&ipm->status, is_write ? IPM_WRITE_BUSY_BIT : IPM_READ_BUSY_BIT);
+	if ((!atomic_test_bit(&ipm->status, IPM_WRITE_BUSY_BIT))
+		&& (!atomic_test_bit(&ipm->status, IPM_READ_BUSY_BIT))) {
+		pm_device_busy_clear(dev);
+	}
+	irq_unlock(key);
+}
+
+static void ipm_event_dispose(IN sedi_ipc_t device, IN uint32_t event,
+			      INOUT void *params)
+{
+	const struct device *dev = (const struct device *)params;
+	struct ipm_sedi_context *ipm = dev->data;
+	uint32_t drbl_in = 0, len;
+
+	LOG_DBG("dev: %u, event: %u", device, event);
+	switch (event) {
+	case SEDI_IPC_EVENT_MSG_IN:
+		if (ipm->rx_msg_notify_cb != NULL) {
+			set_ipm_dev_busy(dev, false);
+			sedi_ipc_read_dbl(device, &drbl_in);
+			len = IPC_HEADER_GET_LENGTH(drbl_in);
+			sedi_ipc_read_msg(device, ipm->incoming_data_buf, len);
+			ipm->rx_msg_notify_cb(dev,
+					      ipm->rx_msg_notify_cb_data,
+					      drbl_in, ipm->incoming_data_buf);
```

Comment:
suggest to just pass the pointer of IPC message registers as the parameter. better than a copied buffer.

Reply:
Will give a new solution in future.